perllib/FixMyStreet/Map/WMSBase.pm

```diff
@@ -0,0 +1,274 @@
+
+    $params{x_left_tile} = $params{x_centre_tile} - int($params{cols} / 2);
+    $params{y_top_tile}  = $params{y_centre_tile} - int($params{rows} / 2);
+
+    $params{pins} = [
+        map {
+            my $pin = { %$_ }; # shallow clone
+            ($pin->{px}, $pin->{py})
+                = $self->latlon_to_px($pin->{latitude}, $pin->{longitude},
+                            @params{'x_left_tile', 'y_top_tile', 'zoom'});
+            $pin;
+        } @{ $params{pins} }
+    ];
+
+    my @scales = $self->scales;
+    return {
+        %params,
+        type => $self->map_template,
+        map_type => 'OpenLayers.Layer.WMS',
+        tiles => $self->map_tiles( %params ),
+        copyright => $self->copyright(),
+        zoom => $params{zoom},
+        zoomOffset => $self->zoom_parameters->{min_zoom_level},
+        numZoomLevels => $self->zoom_parameters->{zoom_levels},
+        tile_size => $self->tile_parameters->{size},
+        tile_dpi => $self->tile_parameters->{dpi},
+        tile_urls => encode_json( $self->tile_parameters->{urls} ),
+        layer_names => encode_json( $self->tile_parameters->{layer_names} ),
+        map_projection => $self->tile_parameters->{projection},
+        wms_version => $self->tile_parameters->{wms_version},
+        format => $self->tile_parameters->{format},
+        scales => encode_json( \@scales ),
+    };
+}
+
+sub tile_base_url {
+    my $self = shift;
+    my $params = $self->tile_parameters;
+    return sprintf '%s?version=%s&format=%s&size=%s&width=%s&height=%s&service=WMS&layers=%s&request=GetMap&srs=%s',
+        $params->{urls}[0], $params->{wms_version}, $params->{format}, $params->{size}, $params->{size},
+        $params->{size}, $params->{layer_names}[0], $params->{projection};
+}
+
+# Given a lat/lon, convert it to OSM tile co-ordinates (nearest actual tile,
+# adjusted so the point will be near the centre of a 2x2 tiled map).
+#
+# Takes parameter for rows/cols.  For even sizes (2x2, 4x4 etc.) will
+# do adjustment, but simply returns actual for odd sizes.
+#
+sub latlon_to_tile_with_adjust {
+    my ($self, $lat, $lon, $zoom, $rows, $cols) = @_;
+    my ($x_tile, $y_tile)
+        = my @ret
+        = $self->reproject_from_latlon($lat, $lon, $zoom);
+
+    my $tile_params = $self->tile_parameters;
+    my $res = $self->get_res($zoom);
+
+    $x_tile = $x_tile -  ($res * $tile_params->{size});
+    $y_tile = $y_tile + ($res * $tile_params->{size});
+
+    return ( int($x_tile), int($y_tile) );
+}
+
+# Given a lat/lon, convert it to pixel co-ordinates from the top left of the map
+sub latlon_to_px($$$$$$) {
+    my ($self, $lat, $lon, $x_tile, $y_tile, $zoom) = @_;
+    my ($pin_x_tile, $pin_y_tile) = $self->reproject_from_latlon($lat, $lon, $zoom);
+    my $res = $self->get_res($zoom);
+    my $pin_x = ( $pin_x_tile - $x_tile ) / $res;
+    my $pin_y = ( $y_tile - $pin_y_tile ) / $res;
+    return ($pin_x, $pin_y);
+}
+
+sub click_to_tile {
```

Comment:
This function appears to be doing what in other map types is two functions, click_to_tile and tile_to_latlon, I don't know if there's a reason for that, I'm guessing the tile X/Y co-ords have already been projected? Just making sure I have it clear in my head :)

Reply:
Yeah, the tile coords can be directly reprojected so there's no need for the second tile_to_latlon function/